### PR TITLE
Firewall Pinhole Fix for ILB and NetLB

### PR DESCRIPTION
**Pull request by sugangli**
/kind bug
Adding new gce firewall field DestinationRanges to ILB and NetLB. This change does not change the existing behaviors while creating these LBs. It simply restricts the access from Node & All IPs to Node & DestinationRanges. 

Two major changes:
1.  Adding new patchFirewall function to support new fields.
2.  Moving firewall creation to the end of the function so that it can use the IP from the forwarding rule.

Before the introduction of GCE pinhole firewall feature, the destination-ranges of the firewall rule were default to all, i.e., any source allowed in the source-ranges was allowed to access any IP in the backend.  This introduced a security risk, for example, any netlb service opens port for 22 would expose ssh access to the node in addition to other security issues. The fix here is aimed at restricting the access to the backend IPs via allowing traffic via ILB or NetLB IP only. 
```release-note
This change picks up the latest GCE pinhole firewall feature, which introduces destination-ranges in the ingress firewall-rules. It restricts the access to the backend IPs via allowing traffic via allowing traffic through ILB or NetLB IP only. This change does NOT change the existing ILB or NetLB behavior. 
```

### CODE REVIEW

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go [hunk 1]
```diff
@@ -225,6 +220,11 @@ func (g *Cloud) ensureInternalLoadBalancer(clusterName, clusterID string, svc *v
 		}
 	}
 
+	// Ensure firewall rules if necessary
+	if err = g.ensureInternalFirewalls(loadBalancerName, ipToUse, clusterID, nm, svc, strconv.Itoa(int(hcPort)), sharedHealthCheck, nodes); err != nil {
```
Comment (kl52752) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:224]: I don't think that it is necessary to query for Forwarding rule inside `ensureInternalFirewalls`. 
Maybe you can just assign ip address from forwarding rule to `ipToUse` if needed and remove the check for empty ipAddress inside the function.

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go [hunk 2]
```diff
@@ -498,15 +502,25 @@ func (g *Cloud) ensureInternalFirewalls(loadBalancerName, ipAddress, clusterID s
 	if err != nil {
 		return err
 	}
-	err = g.ensureInternalFirewall(svc, MakeFirewallName(loadBalancerName), fwDesc, sourceRanges.StringSlice(), portRanges, protocol, nodes, loadBalancerName)
+	var destinationIP string
+	if ipAddress == "" {
```
Comment (kl52752) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:506]: commented above. I think that this is not necessary here

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go (hunk 1 above)
Comment (sugangli) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:229]: Updated.

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go (hunk 2 above)
Comment (sugangli) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go]: Updated.

**Comment by thockin** [comment]
Code-wise I think I get what this is doing.  I don't have the ability to approve in legacy-cloud-providers and I have a hard time keeping track of the state of this whole transition in my head.  @cheftako for consult?

**Comment by sugangli** [comment]
> Code-wise I think I get what this is doing. I don't have the ability to approve in legacy-cloud-providers and I have a hard time keeping track of the state of this whole transition in my head. @cheftako for consult?

Thanks Tim! Will reach out to @cheftako. Actually he pointed me to this repo. :)

**Comment by nckturner** [comment]
It would be helpful for more context in the PR description as to why this is a bug and why its needed (I'm only supposed to approve bugs into legacy providers).  Happy to approve when you add a little more context.

**Comment by sugangli** [comment]
> It would be helpful for more context in the PR description as to why this is a bug and why its needed (I'm only supposed to approve bugs into legacy providers). Happy to approve when you add a little more context.

Thank you for reviewing this PR!  I added some background in the description. Please let me know if I need to provide further detail.

**Comment by nckturner** [comment]
/approve

This is a bug fix.

**Comment by thockin** [comment]
/approve

Who's doing the code-review on this?  I can approve but it would be best to get someone who knows this code well to do a thorogh review.  It all looks OK to me but I didn't go super deep on it.

**Comment by bowei** [comment]
@sugangli -- can we sync for 15 minutes to understand why certain sections of logic was moved around in the function, it will help me review this quickly.

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_external.go [hunk 3]
```diff
@@ -921,7 +915,7 @@ func (g *Cloud) ensureHTTPHealthCheckFirewall(svc *v1.Service, serviceName, ipAd
 			return fmt.Errorf("error getting firewall for health checks: %v", err)
 		}
 		klog.Infof("Creating firewall %v for health checks.", fwName)
-		if err := g.createFirewall(svc, fwName, desc, sourceRanges, ports, hosts); err != nil {
+		if err := g.createFirewall(svc, fwName, desc, "", sourceRanges, ports, hosts); err != nil {
```
Comment (bowei) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_external.go:918]: `g.createFirewall(..., /*destinationIP*/ "", ...`

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_external.go [hunk 4]
```diff
@@ -934,7 +928,7 @@ func (g *Cloud) ensureHTTPHealthCheckFirewall(svc *v1.Service, serviceName, ipAd
 		!equalStringSets(fw.Allowed[0].Ports, []string{strconv.Itoa(int(ports[0].Port))}) ||
 		!equalStringSets(fw.SourceRanges, sourceRanges.StringSlice()) {
 		klog.Warningf("Firewall %v exists but parameters have drifted - updating...", fwName)
-		if err := g.updateFirewall(svc, fwName, desc, sourceRanges, ports, hosts); err != nil {
+		if err := g.updateFirewall(svc, fwName, desc, "", sourceRanges, ports, hosts); err != nil {
```
Comment (bowei) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_external.go:931]: `/*destinationIP*/ ...`

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_external.go [hunk 5]
```diff
@@ -1007,7 +1001,7 @@ func (g *Cloud) updateFirewall(svc *v1.Service, name, desc string, sourceRanges
 	return nil
 }
 
-func (g *Cloud) firewallObject(name, desc string, sourceRanges utilnet.IPNetSet, ports []v1.ServicePort, hosts []*gceInstance) (*compute.Firewall, error) {
+func (g *Cloud) firewallObject(name, desc, destinationIP string, sourceRanges utilnet.IPNetSet, ports []v1.ServicePort, hosts []*gceInstance) (*compute.Firewall, error) {
```
Comment (bowei) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_external.go:1004]: We should document that destinationIP can be empty string "" and this means that it is not set.

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go [hunk 6]
```diff
@@ -416,7 +421,7 @@ func (g *Cloud) teardownInternalHealthCheckAndFirewall(svc *v1.Service, hcName s
 	return nil
 }
 
-func (g *Cloud) ensureInternalFirewall(svc *v1.Service, fwName, fwDesc string, sourceRanges []string, portRanges []string, protocol v1.Protocol, nodes []*v1.Node, legacyFwName string) error {
+func (g *Cloud) ensureInternalFirewall(svc *v1.Service, fwName, fwDesc, DestinationIP string, sourceRanges []string, portRanges []string, protocol v1.Protocol, nodes []*v1.Node, legacyFwName string) error {
```
Comment (bowei) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:424]: don't capitalize DestinationIP

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_external.go (hunk 5 above)
Comment (sugangli) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_external.go:1003]: Added.

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_external.go [hunk 7]
```diff
@@ -123,12 +123,6 @@ func (g *Cloud) ensureExternalLoadBalancer(clusterName string, clusterID string,
 	// keeping the static IP around even though this is more complicated because
 	// it makes it less likely that we'll run into quota issues. Only 7 static
 	// IP addresses are allowed per region by default.
-	//
```
[the remark this answers is not in the document]
Comment (MrHohn) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_external.go:116]: Discussed offline and we decided to move this back so we don't accidentally break implicit logic around this.

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go [hunk 8]
```diff
@@ -225,6 +220,16 @@ func (g *Cloud) ensureInternalLoadBalancer(clusterName, clusterID string, svc *v
 		}
 	}
 
+	fwd, err := g.GetRegionForwardingRule(loadBalancerName, g.region)
```
Comment (MrHohn) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:223]: Pasting comment from offline - merge the second `GetRegionForwardingRule` down below into this so we don't generate additional get call.

**Review by MrHohn — COMMENTED** [review]
Walked through the changes with Sugang offline on this and overall LGTM.

**Review by MrHohn — APPROVED** [review]
Thanks! Could you update the release note since this changes the firewall behavior for new/existing L4 LBs?
/lgtm
/approve

**Pull request merged**